cmd/changelog/cmd/generate.go

```diff
@@ -0,0 +1,300 @@
+			(maxVersion == nil || ver.GreaterThan(maxVersion)) {
+			maxVersion = ver
+			maxRef = tag
+		}
+	}
+	return maxVersion, maxRef, nil
+}
+
+// findNewCommits queries the git repository for new commits since the last tagged release.
+func findNewCommits(r *git.Repository, prevHash plumbing.Hash) ([]Commit, error) {
+	logs, err := r.Log(new(git.LogOptions))
+	if err != nil {
+		return nil, err
+	}
+	defer logs.Close()
+
+	var commits []Commit
+	for {
+		commit, err := logs.Next()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+		if commit.Hash == prevHash {
+			break
+		}
+		c, err := parseCommit(commit)
+		if err != nil {
+			return nil, err
+		}
+		commits = append(commits, c)
+	}
+	return commits, nil
+}
+
+const (
+	Feature = "feat"
+	Fix     = "fix"
+)
+
+// organizeCommits filters commits into groups based on their type.
+func organizeCommits(commits []Commit) (breaking, features, fixes []Commit) {
+	breaking = make([]Commit, 0, len(commits))
+	features = make([]Commit, 0, len(commits))
+	fixes = make([]Commit, 0, len(commits))
+
+	for _, c := range commits {
+		if c.Breaking {
+			breaking = append(breaking, c)
+		}
+		switch c.Type {
+		case Feature:
+			features = append(features, c)
+		case Fix:
+			fixes = append(fixes, c)
+		default:
+			//Ignore commit
+		}
+	}
+	return
+}
+
+// ReleaseType corresponds the semantic version numbers.
+type ReleaseType int
+
+const (
+	Major ReleaseType = iota
+	Minor
+	Patch
+)
+
+// determineNextVersion computes the next semantic version based on the types of commits in the release.
+func determineNextVersion(prevVer *semver.Version, numBreaking, numFeatures int) *semver.Version {
+	typ := Patch
+	if numBreaking > 0 {
+		typ = Major
+		if prevVer.Major() == 0 {
+			// Downgrade breaking change to minor release for pre 1.0.0 releases
+			typ = Minor
+		}
+	} else if numFeatures > 0 {
+		typ = Minor
+	}
+	v := incVersion(typ, *prevVer)
+	return &v
+}
+
+// incVersion increments the semantic version according to the release type.
+func incVersion(typ ReleaseType, v semver.Version) semver.Version {
+	switch typ {
+	case Major:
+		return v.IncMajor()
+	case Minor:
+		return v.IncMinor()
+	case Patch:
+		return v.IncPatch()
+	default:
+		panic("unreachable")
+	}
+}
+
+// Commit represents the conventional commit understanding of a git commit.
+type Commit struct {
+	Type    string
+	Scope   string
+	Subject string
+	Body    string
+	Footer  string
+
+	Breaking bool
+
+	Hash      string
+	ShortHash string
+	Author    string
+}
+
+// headerPattern matches the type, scope and subject of a commit message header.
+var headerPattern = regexp.MustCompile(`^(\w+)(\(\w+\))?:(.*)$`)
+
+const breaking = "BREAKING CHANGE:"
+
+func parseCommit(commit *object.Commit) (Commit, error) {
+	var (
+		header,
+		body,
+		footer string
+	)
+	lines := strings.Split(commit.Message, "\n\n")
+	switch {
+	case len(lines) >= 3:
+		footer = strings.TrimSpace(strings.Join(lines[2:], "\n\n"))
```

Comment:
I hadn't thought too hard about the difference between parsing the body vs the footer.  I don't think there's a simple way to tell when there's just a body with no footer, as opposed to a footer with no body, as opposed to two paragraphs of body with no footer. 

If I'm reading this correctly, it doesn't look like we're using the body or footer anywhere after we set those fields on the Commit. I'd lean towards just merging them as a single field, perhaps `BodyAndFooter`.